### PR TITLE
#10119: move fold op to ttnn infra

### PR DESCRIPTION
**Pull request by mywoodstock**
### Ticket
#10119 

### Problem description
Provide context for the problem.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/test_fold_op.py [hunk 1]
```diff
@@ -123,14 +114,15 @@ def pad_and_fold_with_permute_and_reshape_on_device(
     activation_pyt_padded = ttnn.transpose(activation_pyt_padded, 1, 2, memory_config=ttnn.L1_MEMORY_CONFIG)
     # reshape
     n, w, c, h = activation_pyt_padded.shape
-    activation_pyt_padded = ttl.tensor.reshape(
+    ttnn.tt_lib
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_fold_op.py:117]: remove?

Comment (mywoodstock) [reply tests/ttnn/unit_tests/operations/test_fold_op.py]: removed. not sure how i inserted that!

File: ttnn/cpp/ttnn/operations/data_movement/fold/fold.hpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+#include "ttnn/tensor/host_buffer/functions.hpp"
+
+#include "device/fold_device_op.hpp"
+
+namespace ttnn {
+namespace operations::data_movement {
+
+struct FoldOperation {
+    static ttnn::Tensor operator()(
+        uint8_t queue_id,
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/fold/fold.hpp:22]: we should have a version without queue_id

File: ttnn/cpp/ttnn/operations/data_movement/fold/fold_pybind.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                return op(queue_id, input, stride_h, stride_w, use_transpose_as_fold, output_shape, pad_c, pad_h, pad_w);
+            },
+            py::arg("input"),
+            py::arg("stride_h"),
+            py::arg("stride_w"),
+            py::arg("use_transpose_as_fold") = false,
+            py::arg("output_shape") = std::nullopt,
+            py::arg("pad_c") = 0,
+            py::arg("pad_h") = 0,
+            py::arg("pad_w") = 0,
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/fold/fold_pybind.cpp:42]: should these be kw_only too?

File: ttnn/cpp/ttnn/operations/data_movement/fold/fold_pybind.hpp [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "pybind11/pybind_fwd.hpp"
+
+namespace py = pybind11;
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/fold/fold_pybind.hpp:9]: please remove from the header

File: ttnn/cpp/ttnn/operations/data_movement/fold/fold_pybind.cpp (hunk 3 above)
Comment (mywoodstock) [reply ttnn/cpp/ttnn/operations/data_movement/fold/fold_pybind.cpp:42]: Kept the original behavior.

File: ttnn/cpp/ttnn/operations/data_movement/fold/fold_pybind.hpp (hunk 4 above)
Comment (mywoodstock) [reply ttnn/cpp/ttnn/operations/data_movement/fold/fold_pybind.hpp]: Removed.

File: ttnn/cpp/ttnn/operations/data_movement/fold/fold.hpp (hunk 2 above)
Comment (mywoodstock) [reply ttnn/cpp/ttnn/operations/data_movement/fold/fold.hpp:31]: Added.

**Review by eyonland — APPROVED** [review]

**Review by sjameelTT — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_fold_op.py [hunk 5]
```diff
@@ -123,14 +114,14 @@ def pad_and_fold_with_permute_and_reshape_on_device(
...
-    activation_pyt_padded = ttl.tensor.reshape(
+    activation_pyt_padded = ttnn.tt_lib.tensor.reshape(
         activation_pyt_padded, n, w // stride_w, c * stride_w, h, output_mem_config=ttnn.L1_MEMORY_CONFIG
     )
     # transpose
     activation_pyt_padded = ttnn.transpose(activation_pyt_padded, 2, 3, memory_config=ttnn.L1_MEMORY_CONFIG)
     # reshape
     n, w, h, c = activation_pyt_padded.shape
-    activation_pyt_padded = ttl.tensor.reshape(
+    activation_pyt_padded = ttnn.tt_lib.tensor.reshape(
```
[the remark this answers is not in the document]
Comment (yugaoTT) [reply tests/ttnn/unit_tests/operations/test_fold_op.py:124]: remember it was calling different functions, ttnn::reshape calls tensor.reshape and tt_lib calls reshape op

**Review by ayerofieiev-tt — APPROVED** [review]

**Pull request merged**